### DISCUSSION TITLE
Confusing documentation about tags

### DISCUSSION THREAD

**Discussion by GovindarajanNagarajan-TomTom** [docker/metadata-action#366] [Q&A] [answered]
Hi, 
The Readme, while detailed, IMO, misses the core point of how `metadata-action` deciphers the output tag. I am running into some issues using this action locally in our team's repo and would like some clarifications.

The example in the `README.md` provides the following `tags` definition: 
```
          tags: |
            type=ref,event=branch
            type=ref,event=pr
            type=semver,pattern={{version}}
            type=semver,pattern={{major}}.{{minor}}
            type=sha
```

I am afraid I don't understand this very well. 

1. Is the type of the output tag the `ref` , if the event that triggered this workflow is a `push` to the branch ? 
2. why are there 2 patterns specified for `semver` ? 
```
            type=semver,pattern={{version}}
            type=semver,pattern={{major}}.{{minor}}
```
The docs seem to imply that `semver` is the type of tagging that happens by default when a new tag is created. Is it true ? 
If so,  why are 2 patterns specified ? Are they both used, or is it more of a prioritization over another ? 

And is there an inherent prioritization to the rules defined for the tags ? If so , it is ascending or descending ?

**Comment by crazy-max** [comment, marked as answer]
> 1. Is the type of the output tag the `ref` , if the event that triggered this workflow is a `push` to the branch ?

`ref` tag type is explained here https://github.com/docker/metadata-action#typeref. You also have examples of what to expect in the table below:

> This type handles Git ref (or reference) for the following events:
> 
> * `branch` ; eg. `refs/heads/master`
> * `tag` ; eg. `refs/tags/v1.0.0`
> * `pr` ; eg. `refs/pull/318/merge`
> 
> | Event               | Ref                           | Output           |
> |---------------------|-------------------------------|------------------|
> | `pull_request`      | `refs/pull/2/merge`           | `pr-2`           |
> | `push`              | `refs/heads/master`           | `master`         |
> | `push`              | `refs/heads/my/branch`        | `my-branch`      |
> | `push tag`          | `refs/tags/v1.2.3`            | `v1.2.3`         |
> | `push tag`          | `refs/tags/v2.0.8-beta.67`    | `v2.0.8-beta.67` |
> | `workflow_dispatch` | `refs/heads/master`           | `master`         |

`Ref` is set by GitHub Actions depending on the triggered `Event`.

> 2\. why are there 2 patterns specified for `semver` ?

Also explained in https://github.com/docker/metadata-action#typesemver

`pattern` defaults to Git tag if empty.

> The docs seem to imply that `semver` is the type of tagging that happens by default when a new tag is created. Is it true ?

Yes

> If so, why are 2 patterns specified ? Are they both used, or is it more of a prioritization over another ?

Because you might want to push multiple tags with one being `<major>.<minor>` and another one `<version>`. This is useful to push `<major>.<minor>` tag so user consuming your image can just use `v1.0` without worrying about updating the image ref when patched releases are pushed (`v1.0.1`, `v1.0.2`, ...).
